br/pkg/streamhelper/collector.go

```diff
@@ -0,0 +1,284 @@
+	return sb.String()
+}
+
+func (c *storeCollector) spawn(ctx context.Context) func() (StoreCheckpoints, error) {
+	go c.begin(ctx)
+	return func() (StoreCheckpoints, error) {
+		close(c.input)
+		<-c.doneMessenger
+		if err := c.Err(); err != nil {
+			return StoreCheckpoints{}, err
+		}
+		sc := StoreCheckpoints{
+			HasCheckpoint:    c.checkpoint != 0,
+			Checkpoint:       c.checkpoint,
+			FailureSubranges: c.inconsistent,
+		}
+		return sc, nil
+	}
+}
+
+func (c *storeCollector) flush(ctx context.Context) error {
+	log.Debug("sending batch", zap.Int("size", len(c.currentRequest.Regions)), zap.Uint64("store", c.storeID))
+	cli, err := c.service.GetLogBackupClient(ctx, c.storeID)
+	if err != nil {
+		return err
+	}
+	cps, err := cli.GetLastFlushTSOfRegion(ctx, &c.currentRequest)
+	if err != nil {
+		return err
+	}
+	metrics.GetCheckpointBatchSize.WithLabelValues("checkpoint").Observe(float64(len(c.currentRequest.GetRegions())))
+	c.currentRequest = logbackup.GetLastFlushTSOfRegionRequest{}
+	for _, checkpoint := range cps.Checkpoints {
+		if checkpoint.Err != nil {
+			log.Debug("failed to get region checkpoint", zap.Stringer("err", checkpoint.Err))
+			c.inconsistent = append(c.inconsistent, c.regionMap[checkpoint.Region.Id])
+		} else {
+			if c.onSuccess != nil {
+				c.onSuccess(checkpoint.Checkpoint, c.regionMap[checkpoint.Region.Id])
+			}
+			// assuming the checkpoint would never be zero, use it as the placeholder. (1970 is so far away...)
+			if checkpoint.Checkpoint < c.checkpoint || c.checkpoint == 0 {
+				c.checkpoint = checkpoint.Checkpoint
+			}
+		}
+	}
+	return nil
+}
+
+type runningStoreCollector struct {
+	collector *storeCollector
+	wait      func() (StoreCheckpoints, error)
+}
+
+type clusterCollector struct {
+	mu         sync.Mutex
+	collectors map[uint64]runningStoreCollector
+	noLeaders  []kv.KeyRange
+	onSuccess  onSuccessHook
+
+	// The context for spawning sub collectors.
+	// Because the collectors are running lazily,
+	// keep the initial context for all subsquent goroutines,
+	// so we can make sure we can cancel all subtasks.
+	masterCtx context.Context
+	cancel    context.CancelFunc
+	srv       LogBackupService
+}
+
+func NewClusterCollector(ctx context.Context, srv LogBackupService) *clusterCollector {
+	cx, cancel := context.WithCancel(ctx)
+	return &clusterCollector{
+		collectors: map[uint64]runningStoreCollector{},
+		masterCtx:  cx,
+		cancel:     cancel,
+		srv:        srv,
+	}
+}
+
+func (c *clusterCollector) setOnSuccessHook(hook onSuccessHook) {
+	c.onSuccess = hook
+}
+
+func (c *clusterCollector) collectRegion(r RegionWithLeader) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.masterCtx.Err() != nil {
+		return nil
+	}
+
+	if r.Leader.GetStoreId() == 0 {
+		log.Warn("there is regions without leader", zap.Uint64("region", r.Region.GetId()))
+		c.noLeaders = append(c.noLeaders, kv.KeyRange{StartKey: r.Region.StartKey, EndKey: r.Region.EndKey})
```

Comment:
need return

Reply:
and another thing is `noLeaders` seems always increase and never clear.